## Setup — load the data

In [8]:
import numpy as np

# Columns: 0 day | 1 date | 2 temp_c | 3 rainfall_mm | 4 humidity_pct
dates = np.genfromtxt("week5_kigali_weather.csv", delimiter=",", skip_header=1, usecols=1, dtype=str)
temps = np.genfromtxt("week5_kigali_weather.csv", delimiter=",", skip_header=1, usecols=2)
rain  = np.genfromtxt("week5_kigali_weather.csv", delimiter=",", skip_header=1, usecols=3)
hum   = np.genfromtxt("week5_kigali_weather.csv", delimiter=",", skip_header=1, usecols=4)

print("Records loaded:", len(dates))
print("First date:", dates[0], "| Last date:", dates[-1])

Records loaded: 60
First date: 2026-09-01 | Last date: 2026-10-30


---
# Part 1 · List Comprehensions (6 marks)

In [9]:
temps_list = [24.4, 25.9, 28.6, 19.3, 26.0, 22.1]   # given for Part 1
raw = ["24.4", "25.9", "28.6", "19.3"]              # given for Ex 1.3

### Exercise 1.1 — Transform (2 marks)
Convert every temperature to Fahrenheit with `t * 9/5 + 32`, in one line.

In [10]:
temps_f = [t * 9/5 + 32 for t in temps_list]
print(temps_f)

[75.92, 78.62, 83.48, 66.74000000000001, 78.8, 71.78]


### Exercise 1.2 — Filter (2 marks)

In [11]:
# (a) hot = temperatures >= 26
hot = [t for t in temps_list if t >= 26]
print("hot:", hot)

# (b) cool_rounded = temperatures below 24, rounded with round(t)
cool_rounded = [round(t) for t in temps_list if t < 24]
print("cool_rounded:", cool_rounded)

hot: [28.6, 26.0]
cool_rounded: [19, 22]


### Exercise 1.3 — Convert types (2 marks)
CSV files give you strings. Convert `raw` to floats, then print the average.

In [12]:
raw_floats = [float(x) for x in raw]
print("floats:", raw_floats)
print("average:", sum(raw_floats) / len(raw_floats))

floats: [24.4, 25.9, 28.6, 19.3]
average: 24.55


---
# Part 2 · lambda (4 marks)

### Exercise 2.1 — Rank with `sorted` (2 marks)

In [6]:
results = [("Aline", 85), ("Eric", 78), ("Diane", 91), ("Samuel", 92), ("Grace", 76)]

ranked = sorted(results, key=lambda r: r[1], reverse=True)   # r[1] is the score
print("Top 3:", ranked[:3])

Top 3: [('Samuel', 92), ('Diane', 91), ('Aline', 85)]


### Exercise 2.2 — `max` with a key (2 marks)

In [7]:
rain_log = [(3, 0.0), (7, 12.5), (11, 3.2), (14, 26.2), (18, 8.9)]   # (day, rainfall_mm)

wettest = max(rain_log, key=lambda pair: pair[1])
print("Wettest day:", wettest)

Wettest day: (14, 26.2)


---
# Part 3 · NumPy Arrays (8 marks)

### Exercise 3.1 — The speed test (2 marks)
This is the provided speed-test cell: a plain Python loop versus `arr * 2` on one million numbers. (It is the one place a loop is expected here — it is the thing we are comparing against.)

In [3]:
import time

N = 1_000_000
py_list = list(range(N))
arr = np.arange(N)

# Python loop
start = time.time()
doubled_list = []
for x in py_list:
    doubled_list.append(x * 2)
loop_time = time.time() - start

# NumPy vectorized
start = time.time()
doubled_arr = arr * 2
numpy_time = time.time() - start

print(f"Python loop : {loop_time:.5f} s")
print(f"NumPy       : {numpy_time:.5f} s")
print(f"NumPy was about {loop_time / numpy_time:.0f}x faster")

# Comment: on my machine NumPy was roughly 15x faster than the Python loop (re-run to see your own number).

Python loop : 0.24082 s
NumPy       : 0.00390 s
NumPy was about 62x faster


### Exercise 3.2 — Describe the temperatures (3 marks)
Array methods only. **Checkpoint:** the mean should be 24.00 °C.

In [9]:
print(f"Mean temperature : {temps.mean():.2f} °C")
print(f"Max temperature  : {temps.max():.2f} °C")
print(f"Min temperature  : {temps.min():.2f} °C")
print(f"Std deviation    : {temps.std():.2f} °C")

Mean temperature : 24.00 °C
Max temperature  : 28.60 °C
Min temperature  : 19.30 °C
Std deviation    : 2.61 °C


### Exercise 3.3 — The hottest day (3 marks)

In [10]:
position = np.argmax(temps)
print(f"Hottest day: {dates[position]} with {temps[position]} °C")

Hottest day: 2026-09-17 with 28.6 °C


---
# Part 4 · Boolean Masking (10 marks)

### Exercise 4.1 — Hot days (3 marks)
A hot day is ≥ 26 °C.

In [13]:
hot_mask = temps >= 26

print("(a) Number of hot days:", hot_mask.sum())          # True counts as 1

print(f"(b) Average temperature on hot days: {temps[hot_mask].mean():.2f} °C")

(a) Number of hot days: 18
(b) Average temperature on hot days: 27.11 °C


### Exercise 4.2 — Rain report (3 marks)

In [14]:
rainy_mask = rain > 0

print("(a) Number of rainy days:", rainy_mask.sum())
print(f"(b) Total rainfall over 60 days: {rain.sum():.1f} mm")
print(f"(c) Heaviest single-day rainfall: {rain.max():.1f} mm")

(a) Number of rainy days: 16
(b) Total rainfall over 60 days: 331.2 mm
(c) Heaviest single-day rainfall: 26.2 mm


### Exercise 4.3 — Cross-array analysis (4 marks)

In [13]:
# (a) Average temperature: rainy days vs dry days (mask from rain, data from temps)
print(f"(a) Rainy days avg temp: {temps[rain > 0].mean():.2f} °C")
print(f"    Dry days avg temp  : {temps[rain == 0].mean():.2f} °C")

# (b) Days that were hot AND rainy (parentheses around each condition!)
hot_and_rainy = (temps >= 26) & (rain > 0)
print("(b) Hot AND rainy days:", hot_and_rainy.sum())

# (c) Days with humidity >= 70
print("(c) Days with humidity >= 70:", (hum >= 70).sum())

(a) Rainy days avg temp: 24.14 °C
    Dry days avg temp  : 23.95 °C
(b) Hot AND rainy days: 5
(c) Days with humidity >= 70: 16


---
# Part 5 · Anomalies & Trends (7 marks)

### Exercise 5.1 — Anomaly detection (4 marks)

In [15]:
anomaly = temps - temps.mean()      # vectorized: one subtraction for all 60 days

# (a) DATES of days more than 3 °C above the mean
print("(a) Days more than 3 °C above the mean:", dates[anomaly > 3])

# (b) Count of days more than 3 °C below the mean
print("(b) Days more than 3 °C below the mean:", (anomaly < -3).sum())

(a) Days more than 3 °C above the mean: ['2026-09-07' '2026-09-09' '2026-09-11' '2026-09-12' '2026-09-14'
 '2026-09-16' '2026-09-17' '2026-09-19' '2026-09-22']
(b) Days more than 3 °C below the mean: 8


### Exercise 5.2 — Is the season changing? (3 marks)

In [16]:
first_half = temps[:30].mean()
last_half  = temps[30:].mean()

print(f"First 30 days average: {first_half:.2f} °C")
print(f"Last 30 days average : {last_half:.2f} °C")
print(f"Difference           : {last_half - first_half:+.2f} °C")

# Comment: Kigali got COOLER. The average fell from about 26.2 °C in the first 30 days
# to about 21.8 °C in the last 30 days, a drop of roughly 4.4 °C.

First 30 days average: 26.21 °C
Last 30 days average : 21.78 °C
Difference           : -4.43 °C


---
# Bonus (+3) — Top 5 wettest days
Pairing dates with rainfall using `zip`, ranking with `sorted(..., key=lambda ...)`. A loop is allowed here only for printing.

In [17]:
wettest5 = sorted(zip(dates, rain), key=lambda pair: pair[1], reverse=True)[:5]

for date, mm in wettest5:
    print(f"{date}  ->  {mm} mm")

2026-09-07  ->  26.2 mm
2026-09-14  ->  25.1 mm
2026-09-21  ->  24.0 mm
2026-09-05  ->  22.9 mm
2026-09-28  ->  22.9 mm


---
# Part 6 · Reflection



## 1. What did NumPy hide from me, and why was it still important to learn the loop version first?

In Week 3 I averaged a column by hand: make a `total = 0`, loop over every row, add each value, count the rows, then divide. `temps.mean()` does all of that in one call, and it runs the loop in fast compiled code instead of slow Python. NumPy hid the walking-through-the-data step, the accumulator variable, and the counting.

Learning the loop first was still important because it showed me what "average" actually does. Without that, `temps.mean()` would be a magic word. Now I can predict what a method will do, spot mistakes (for example `std` and `mean` on the wrong array), and write a loop myself when no ready-made method exists. The loop version also made the speed test mean something: I saw how much work was being skipped.

## 2. One insight about Kigali's weather##

Kigali got clearly cooler over the 60 days. The average temperature fell from 26.21 °C in the first 30 days to 21.78 °C in the last 30, a drop of 4.43 °C. All 18 hot days (26 °C or above) happened in the first half and none in the second. At the same time, humidity rose (about 58% to 66%), and the heaviest rain also eased (26.2 mm on 7 September, but at most 21.8 mm in the second half). Rainy days were equal in both halves (8 and 8), but total rainfall fell from 183 mm to 148 mm.